packages/salesforcedx-vscode-core/src/deploys/toolingDeploy.ts

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { AuthInfo, Connection } from '@salesforce/core';
+import * as fs from 'fs';
+import * as path from 'path';
+import {
+  supportedToolingTypes,
+  ToolingCreateResult,
+  ToolingRetrieveResult
+} from './index';
+// tslint:disable-next-line:no-var-requires
+const DOMParser = require('xmldom-sfdx-encoding').DOMParser;
+
+export class ToolingDeploy {
+  private static instance: ToolingDeploy;
+  private metadataType?: string;
+  private connection?: Connection;
+  private apiVersion?: string;
+
+  private container?: ToolingCreateResult;
+  private containerMember?: ToolingCreateResult;
+  private asyncRequest?: ToolingCreateResult;
+
+  public constructor() {}
+
+  public static getInstance(): ToolingDeploy {
+    if (!this.instance) {
+      this.instance = new ToolingDeploy();
+    }
+    return this.instance;
+  }
+
+  public async init(username: string, apiVersion?: string) {
+    try {
+      const connection = await Connection.create({
+        authInfo: await AuthInfo.create({ username })
+      });
+      this.connection = connection;
+      if (apiVersion) this.apiVersion = apiVersion;
+    } catch (e) {
+      throw e;
+    }
+  }
+
+  public async deploy(options: any) {
```

Comment:
We should have the options parameter typed so consumers of the eventual library have useful autocompletion and there's type checking, considering it's a primary endpoint

Reply:
Oops yes, good point I thought I'd covered that 😅